Update routing and navigation guide to router 0.10.x

klequis opened the pull request:
I made some significant changes. I'm totally open to making any changes in structure and topics, just let me know. If you are happy with it, it is ready to go.


I based this revision off of solid router's README in its repo. Topics were mostly the same as the original content.

All content is now updated to solid router 10.x

Since "guides" are under "learn" in the doc, I made the assumption that they should be more example based then the "reference" stuff, so I expanded a couple of examples, most notably Load Functions, which I initially struggled to understand.

The first section I called "Routing Quick Start". It is a code complete example of basic routing.

## Removed suff
To reduce duplication and keep things brief I removed the below sections and added links to them at the bottom of the page:
- Components
- Migrations from 0.9.x
- Data APIs - this section is relevant but is completely duplicative of what is in "Reference". I'm happy to put it back if you think that is best.

I also removed the Final callout that directed readers to the router's readme, as the docs now contain all the information contained there.


## Added links at end of page to

- Solid Router Components 
- Solid Router Data APIs
- Migration from 0.9.x

LadyBluenotes reviewed (changes requested):
Thank you for looking through this! However, before I get to reviewing it can you make sure that every sentence is it's own line and that you've reviewed the [WRITING.md file](https://github.com/solidjs/solid-docs-next/blob/main/WRITING.md)? There are some things in there to take into account prior to a proper review (such as avoiding the use of "we").

klequis commented:
Updated

- one line per sentence
- Paragraphs have 2 lines between them
- Corrected several pronouns to second person
- Add more internal links. Where the content was the same in the guide and reference I did not add a link.

LadyBluenotes reviewed (changes requested):
This looks like a lot 😅 but thank you so much for taking on this guide! I really appreciate it. Let me know if you have any questions, comments or concerns regarding what I did :)

LadyBluenotes reviewed (commented):
This looks like a lot 😅 but thank you so much for taking on this guide! I really appreciate it. Let me know if you have any questions, comments or concerns regarding what I did :)

klequis commented:
This is my first time I have had changes reviewed on a pull request. Not that my changes are so good, I just haven't done many pull requests :).

The missing section on useParams was an oversight. It is definitely needed.

I planned to accept all of your changes and then add the userParams section back. Unfortunately, the review interface doesn't currently work for  "deleted lines", so anywhere you deleted a line the changes is still outstanding.

The easiest way for me to finish-up is to see your suggestions as a complete document. I don't know how to do that but I'll look into it in the morning. Suggestions are welcome .

klequis commented on src/routes/guides/routing-and-navigation.mdx:
```diff
@@ -21,663 +23,734 @@ npm install @solidjs/router
 
-When working with dynamic routes, it is often necessary to validate the route parameters to ensure they meet certain criteria.
-This task can be performed through `MatchFilters`.
 
```
I don't see this in the preview anymore. Marking as resolved.

klequis commented:
I hope my lack of pull review knowledge didn't cause you extra work or delayed this getting done. I think I have figured out how things work now.

## 7 changes are outstanding

I don't know how changes get marked as "Outdated" nor the best way to get around that. One solution might be to accept and merge all the changes. I can then quickly make the 7 outstanding changes in a new pull request.

1. https://github.com/solidjs/solid-docs-next/pull/588/files/166237c4c1ac2d88f7356c322939f793493d1b1d#r1529015940

This change is marked as "outdated" and therefore cannot be committed from the PR.

2. https://github.com/solidjs/solid-docs-next/pull/588/files#r1529005068

I do see now the syntax for putting the file name in a code example is 
```

```jsx title="[id].jsx"

```
But when I try to commit this change I get the error message: "Applying suggestions on deleted lines is currently not supported."


3. https://github.com/solidjs/solid-docs-next/pull/588/files/166237c4c1ac2d88f7356c322939f793493d1b1d#r1528983981

Also marked as outdated.

4. https://github.com/solidjs/solid-docs-next/pull/588/files/166237c4c1ac2d88f7356c322939f793493d1b1d#r1528972133

Marked as outdated

Regarding "waterfalls", I don't know what is meant. The sentence comes from `solidjs/router` README. What if we just left that sentence out?


5. https://github.com/solidjs/solid-docs-next/pull/588/files/166237c4c1ac2d88f7356c322939f793493d1b1d#r1528966350 

Marked as outdated

6. https://github.com/solidjs/solid-docs-next/pull/588/files/166237c4c1ac2d88f7356c322939f793493d1b1d#r1528957271

Marked as outdated

7. https://github.com/solidjs/solid-docs-next/pull/588/files#r1528920084

Is an oversight on my part. This section does need to be added back in. I think it will be easy to do once the above 6 outstanding items are resolved.

klequis commented:
The best way to approach this may be for me to consolidate all the changes into a new pull request. I'll have time to do that tomorrow or Monday.

klequis commented:
I still need to review the changes I merged in earlier today. I should have time later today to do that.

klequis commented:
current: The colon that indicates that `id` can be any string, and as long as the URL fits that pattern, 
the `<User>` component will show.

I'm thinking about this one. I don't remember if I wrote it or not:
- Does "The colon indicate that `id` can be any string"? OR, does the colon indicate that `id` is a parameter? I think the latter. 
- `id` 'can be any string' because it is a parameter name

So maybe:

"The colon indicates `id` is a parameter name, which can be any arbitrary string."

(I'v never been fond of the word "arbitrary", but but is part of programmers jargon so useful here.)

klequis commented:
- Are we missing an important point by not mentioning the hover-over functionality.

old: With load functions, data fetching is started parallel to loading the route so it can be used as soon as possible.

new: With load functions, data fetching is started parallel to loading the route, or when a link to the route is hovered over, so it can be used as soon as possible.

klequis commented:
All my changes are in now :)

LadyBluenotes commented:
Thank you for all your hard work @klequis !
